core/src/jsMain/kotlin/Peripheral.kt

```diff
@@ -54,8 +57,36 @@ public class JsPeripheral internal constructor(
     public override val services: List<DiscoveredService>?
         get() = platformServices?.map { it.toDiscoveredService() }
 
-    public override suspend fun rssi(): Int {
-        TODO("Not yet implemented")
+    private val supportsAdvertisements = js("BluetoothDevice.prototype.watchAdvertisements") != null
+
+    public override suspend fun rssi(): Int = suspendCancellableCoroutine { continuation ->
+        check(supportsAdvertisements) { "watchAdvertisements unavailable" }
+
+        var listener: ((JsEvent) -> Unit)? = null
+        val cleanup = {
+            bluetoothDevice.removeEventListener(ADVERTISEMENT_RECEIVED, listener)
+            // At the time of writing `unwatchAdvertisements()` remains unimplemented
+            if (bluetoothDevice.watchingAdvertisements && js("BluetoothDevice.prototype.unwatchAdvertisements") != null) {
+                bluetoothDevice.unwatchAdvertisements()
+            }
+        }
+
+        listener = {
+            val event = it as BluetoothAdvertisingEvent
+            cleanup()
+            if (continuation.isActive) {
+                continuation.resume(event.rssi, null)
+            }
```

Comment:
There shouldn't be any need to check `isActive`.

Per the docs:

> Resumes this continuation with the specified `value` and calls the specified `onCancellation`
handler when either resumed too late (when continuation was already cancelled) or, although resumed
successfully (before cancellation), the coroutine's job was cancelled before it had a
chance to run in its dispatcher, so that the suspended function threw an exception
instead of returning this value.

So, I believe, that it will silently ignore call to `resume` if already cancelled.

```suggestion
            continuation.resume(event.rssi, onCancellation = null)
```

If `onCancellation` handler had been specified then it would be called if we called `resume` in a cancelled (`!isActive`) state. Since we hold the reference to the resource, we still clean up (so no need for us to use `onCancellation`).

The documentation gives the following example of when `onCancellation` is useful (when we're returned/resuming with a resource that will eventually need clean up — `onCancellation` can prevent leaks):

```kotlin
continuation.resume(resource) {
    resource.close()
}
```

Reply:
I read elsewhere in the docs (which I can't find now...) that `resume` throws an exception when called in the inactive state? Given that this callback can repeat I went with the defensive approach. It should never happen as we always remove the listener anyhow so is *probably* safe to ignore but "should never happen" has a way of happening when one least expects it...

Reply:
https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/CancellableContinuationImpl.kt#L265

Reply:
Hmm, the code you link implies that it throws when a `continuation` has already been resumed (not necessarily when inactive). We shouldn't be attempted subsequent resumes, so I don't think we'd be hit by this particular exception being thrown.

It's quite possible we do need the check you have, I just interpreted the documentation that we don't need it.
The documentation has also been wrong in the past.

🤷

Reply:
Spoke on Zoom, makes sense to keep `isActive` check.